{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pysam\n",
    "import re\n",
    "import argparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pairs(regions, chrom1, chrom2):\n",
    "    \"\"\"\n",
    "    Fetch all paired reads, including those with split alignments\n",
    "\n",
    "    Parameters:\n",
    "    - regions: the two SV regions to collect reads from\n",
    "    - chrom1: chromosome # of first region\n",
    "    - chrom2: chromosome # of second region\n",
    "\n",
    "    Returns:\n",
    "    - split_alignments: contains all sets of paired reads with split alignments\n",
    "    - paired_alignments: contains all sets of paired reads without split alignments\n",
    "    \"\"\"\n",
    "    split_alignments = []\n",
    "    nonsplit_alignments = []\n",
    "\n",
    "    # Iterate over first breakpoint end\n",
    "    for i, region in enumerate(regions):\n",
    "        for aln in region:\n",
    "            # If we've seen this read before, skip it\n",
    "            if any(aln in tup for tup in split_alignments) or any(aln in tup for tup in nonsplit_alignments):\n",
    "                # print(\"Duplicate found\")\n",
    "                continue\n",
    "\n",
    "            mate = samfile.mate(aln)\n",
    "\n",
    "            if aln.is_supplementary or aln.has_tag(\"SA\"):\n",
    "                sa_tag = (aln.get_tag(\"SA\")).split(\";\")[:-1]\n",
    "                sa_tag = sa_tag[0]\n",
    "                sup_name, sup_pos, sup_strand, sup_cigar, sup_mapq, sup_tlen = sa_tag.split(\",\")\n",
    "                sup_pos = int(sup_pos)\n",
    "                sup_tlen = int(sup_tlen)\n",
    "\n",
    "                found_split2 = False\n",
    "                for sup_read in samfile.fetch(sup_name, sup_pos, sup_pos + 1):\n",
    "                    if (sup_read.query_name == aln.query_name and sup_read.reference_name == sup_name and (sup_read.reference_start + 1) == sup_pos and re.sub(r'[SH]', '', sup_read.cigarstring) == re.sub(r'[SH]', '', sup_cigar)):\n",
    "                        non_sup = aln if not aln.is_supplementary else sup_read\n",
    "                        sup = aln if aln.is_supplementary else sup_read\n",
    "                        read1 = non_sup if non_sup.is_read1 else mate\n",
    "                        read2 = non_sup if non_sup.is_read2 else mate\n",
    "                        split_alignments.append((read1, sup, read2))\n",
    "                        found_split2 = True\n",
    "                        break\n",
    "                if not found_split2:\n",
    "                    # print(sup_name, sup_pos, sup_cigar, sup_tlen)\n",
    "                    raise NameError(\"split mate not found\")\n",
    "\n",
    "            elif mate.is_supplementary or mate.has_tag(\"SA\"):\n",
    "                sa_tag = (mate.get_tag(\"SA\")).split(\";\")[:-1]\n",
    "                sa_tag = sa_tag[0]\n",
    "                sup_name, sup_pos, sup_strand, sup_cigar, sup_mapq, sup_tlen = sa_tag.split(\",\")\n",
    "                sup_pos = int(sup_pos)\n",
    "                sup_tlen = int(sup_tlen)\n",
    "\n",
    "                found_split2 = False\n",
    "                for sup_read in samfile.fetch(sup_name, sup_pos, sup_pos + 1):\n",
    "                    if (sup_read.query_name == mate.query_name and sup_read.reference_name == sup_name and (sup_read.reference_start + 1) == sup_pos and re.sub(r'[SH]', '', sup_read.cigarstring) == re.sub(r'[SH]', '', sup_cigar)):\n",
    "                        non_sup = mate if not mate.is_supplementary else sup_read\n",
    "                        sup = mate if mate.is_supplementary else sup_read\n",
    "                        read1 = non_sup if non_sup.is_read1 else aln\n",
    "                        read2 = non_sup if non_sup.is_read2 else aln\n",
    "                        split_alignments.append((read1, sup, read2))\n",
    "                        found_split2 = True\n",
    "                        break\n",
    "                if not found_split2:\n",
    "                    # print(sup_name, sup_pos, sup_cigar, sup_tlen)\n",
    "                    raise NameError(\"split mate not found\")\n",
    "                \n",
    "            elif not aln.is_proper_pair and aln.next_reference_name == (chrom2 if i == 0 else chrom1):\n",
    "                read1 = aln if aln.is_read1 else mate\n",
    "                read2 = aln if aln.is_read2 else mate\n",
    "                nonsplit_alignments.append((read1, read2))\n",
    "            \n",
    "            elif aln.is_proper_pair:\n",
    "                read1 = aln if aln.is_read1 else mate\n",
    "                read2 = aln if aln.is_read2 else mate\n",
    "                nonsplit_alignments.append((read1, read2))\n",
    "    \n",
    "    return split_alignments, nonsplit_alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_alignments(bamfile, chrom1, pos1, chrom2, pos2, sv_type, read_support, features, orientation, hom_len, hom, region_size=150, mapq_threshold=15):\n",
    "    \"\"\"\n",
    "    Fetch reads from a BAM file that have split alignments \n",
    "    between two genomic coordinates as well as all paired-end reads.\n",
    "\n",
    "    Parameters:\n",
    "    - bamfile: str, path to the BAM file.\n",
    "    - chrom1: str, chromosome for the first position.\n",
    "    - pos1: int, first position in the genome.\n",
    "    - chrom2: str, chromosome for the second position.\n",
    "    - pos2: int, second position in the genome.\n",
    "    - region_size: int, the size of the region around the positions to fetch alignments from.\n",
    "    - mapq_threshold: int, cutoff for collecting reads above certain mapping quality.\n",
    "\n",
    "    Returns:\n",
    "    - split_df: DataFrame, details of split alignments.\n",
    "    - paired_df: DataFrame, details of discordant alignments.\n",
    "    \"\"\"\n",
    "    # Open the BAM file\n",
    "    samfile = pysam.AlignmentFile(bamfile, \"rb\")\n",
    "\n",
    "    # Fetch alignments region_size away from both positions\n",
    "    region1 = samfile.fetch(chrom1, pos1 - region_size, pos1 + region_size + 1)\n",
    "    region2 = samfile.fetch(chrom2, pos2 - region_size, pos2 + region_size + 1)\n",
    "\n",
    "    split_alignments, nonsplit_alignments = get_pairs([region1, region2], chrom1, chrom2)\n",
    "    # Collect details of split alignments\n",
    "    split_alignment_details = []\n",
    "    for pair in split_alignments:\n",
    "        read1_1, read1_2, read2 = pair\n",
    "\n",
    "        # Filter out low quality reads\n",
    "        if  ((read1_1.mapping_quality > mapq_threshold and read1_2.mapping_quality > mapq_threshold) or read2.mapping_quality > mapq_threshold) or ((read1_1.is_mapped and read1_2.is_mapped) or read2.is_mapped):\n",
    "            for read in pair:\n",
    "                split_alignment_details.append({\n",
    "                    \"break_chrom1\": chrom1,\n",
    "                    \"break_pos1\": pos1,\n",
    "                    \"break_chrom2\": chrom2,\n",
    "                    \"break_pos2\": pos2,\n",
    "                    \"break_sv_type\": sv_type,\n",
    "                    \"break_read_support\": read_support,\n",
    "                    \"break_features\": features,\n",
    "                    \"break_orientation\": orientation,\n",
    "                    \"AA_homology_len\": hom_len,\n",
    "                    \"AA_homology_seq\": hom,\n",
    "                    \"query_name\": read.query_name,\n",
    "                    \"query_short\": read.query_name.split(':')[-1],\n",
    "                    \"split\": read.has_tag(\"SA\"),\n",
    "                    \"proper_pair\": \"Concordant\" if read.is_proper_pair else \"Discordant\",\n",
    "                    \"read_num\": \"1\" if read.is_read1 else \"2\",\n",
    "                    \"query_chrom\": read.reference_name,\n",
    "                    \"query_pos\": read.reference_start + 1,\n",
    "                    \"query_end\": read.reference_end + 1,\n",
    "                    \"query_orientation\": \"+\" if read.is_forward else \"-\",\n",
    "                    \"query_cigar\": read.cigarstring,\n",
    "                    \"query_aln_full\": read.query_sequence,\n",
    "                    \"query_aln_sub\": read.query_alignment_sequence\n",
    "                })\n",
    "\n",
    "    # Collect details of nonsplit alignments\n",
    "    nonsplit_alignment_details = []\n",
    "    for pair in nonsplit_alignments:\n",
    "        read1, read2 = pair\n",
    "        # Filter out low quality reads\n",
    "        if  (read1.mapping_quality > mapq_threshold or read2.mapping_quality > mapq_threshold) or (read1.is_mapped or read2.is_mapped):\n",
    "            for read in pair:\n",
    "                nonsplit_alignment_details.append({\n",
    "                    \"break_chrom1\": chrom1,\n",
    "                    \"break_pos1\": pos1,\n",
    "                    \"break_chrom2\": chrom2,\n",
    "                    \"break_pos2\": pos2,\n",
    "                    \"break_sv_type\": sv_type,\n",
    "                    \"break_read_support\": read_support,\n",
    "                    \"break_features\": features,\n",
    "                    \"break_orientation\": orientation,\n",
    "                    \"AA_homology_len\": hom_len,\n",
    "                    \"AA_homology_seq\": hom,\n",
    "                    \"query_name\": read.query_name,\n",
    "                    \"query_short\": read.query_name.split(':')[-1],\n",
    "                    \"split\": read.has_tag(\"SA\"),\n",
    "                    \"proper_pair\": \"Concordant\" if read1.is_proper_pair else \"Discordant\",\n",
    "                    \"read_num\": \"1\" if read.is_read1 else \"2\",\n",
    "                    \"query_chrom\": read.reference_name,\n",
    "                    \"query_pos\": read.reference_start + 1,\n",
    "                    \"query_end\": read.reference_end + 1,\n",
    "                    \"query_orientation\": \"+\" if read.is_forward else \"-\",\n",
    "                    \"query_cigar\": read.cigarstring,\n",
    "                    \"query_aln_full\": read.query_sequence,\n",
    "                    \"query_aln_sub\": read.query_alignment_sequence\n",
    "                })\n",
    "        \n",
    "    samfile.close()\n",
    "\n",
    "    # Convert details to pandas DataFrames\n",
    "    split_df = pd.DataFrame(split_alignment_details)\n",
    "    nonsplit_df = pd.DataFrame(nonsplit_alignment_details)\n",
    "    # paired_df = pd.DataFrame(paired_alignment_details)\n",
    "\n",
    "    return split_df, nonsplit_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom1</th>\n",
       "      <th>pos1</th>\n",
       "      <th>chrom2</th>\n",
       "      <th>pos2</th>\n",
       "      <th>sv_type</th>\n",
       "      <th>read_support</th>\n",
       "      <th>features</th>\n",
       "      <th>orientation</th>\n",
       "      <th>homology_length</th>\n",
       "      <th>homology_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9</td>\n",
       "      <td>134155523</td>\n",
       "      <td>chr13</td>\n",
       "      <td>108661410</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>6</td>\n",
       "      <td>unknown</td>\n",
       "      <td>++</td>\n",
       "      <td>2.0</td>\n",
       "      <td>AT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>81088032</td>\n",
       "      <td>chr13</td>\n",
       "      <td>81469934</td>\n",
       "      <td>inversion</td>\n",
       "      <td>4</td>\n",
       "      <td>ecDNA</td>\n",
       "      <td>--</td>\n",
       "      <td>3.0</td>\n",
       "      <td>AAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr13</td>\n",
       "      <td>81109413</td>\n",
       "      <td>chr13</td>\n",
       "      <td>90438706</td>\n",
       "      <td>deletion-like</td>\n",
       "      <td>3</td>\n",
       "      <td>ecDNA</td>\n",
       "      <td>+-</td>\n",
       "      <td>3.0</td>\n",
       "      <td>TGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr13</td>\n",
       "      <td>81470035</td>\n",
       "      <td>chr13</td>\n",
       "      <td>81470310</td>\n",
       "      <td>foldback</td>\n",
       "      <td>3</td>\n",
       "      <td>ecDNA</td>\n",
       "      <td>++</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ATAAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr13</td>\n",
       "      <td>90862773</td>\n",
       "      <td>chr13</td>\n",
       "      <td>90864806</td>\n",
       "      <td>deletion-like</td>\n",
       "      <td>3</td>\n",
       "      <td>ecDNA</td>\n",
       "      <td>+-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr13</td>\n",
       "      <td>92474010</td>\n",
       "      <td>chr13</td>\n",
       "      <td>92475757</td>\n",
       "      <td>foldback</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>++</td>\n",
       "      <td>2.0</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr13</td>\n",
       "      <td>94023407</td>\n",
       "      <td>chr13</td>\n",
       "      <td>108500972</td>\n",
       "      <td>deletion-like</td>\n",
       "      <td>3</td>\n",
       "      <td>unknown</td>\n",
       "      <td>+-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr22</td>\n",
       "      <td>18953122</td>\n",
       "      <td>chr22</td>\n",
       "      <td>22935373</td>\n",
       "      <td>inversion</td>\n",
       "      <td>5</td>\n",
       "      <td>unknown</td>\n",
       "      <td>--</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr22</td>\n",
       "      <td>20950573</td>\n",
       "      <td>chr22</td>\n",
       "      <td>20953841</td>\n",
       "      <td>deletion-like</td>\n",
       "      <td>4</td>\n",
       "      <td>unknown</td>\n",
       "      <td>+-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr22</td>\n",
       "      <td>23049707</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23108069</td>\n",
       "      <td>deletion-like</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>+-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr22</td>\n",
       "      <td>23049910</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23107858</td>\n",
       "      <td>duplication-like</td>\n",
       "      <td>2</td>\n",
       "      <td>ecDNA</td>\n",
       "      <td>-+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom1       pos1 chrom2       pos2           sv_type  read_support  \\\n",
       "0    chr9  133607146  chr22   23632741  interchromosomal             7   \n",
       "1    chr9  134155523  chr13  108661410  interchromosomal             6   \n",
       "2   chr13   81088032  chr13   81469934         inversion             4   \n",
       "3   chr13   81109413  chr13   90438706     deletion-like             3   \n",
       "4   chr13   81470035  chr13   81470310          foldback             3   \n",
       "5   chr13   90862773  chr13   90864806     deletion-like             3   \n",
       "6   chr13   92474010  chr13   92475757          foldback             3   \n",
       "7   chr13   94023407  chr13  108500972     deletion-like             3   \n",
       "8   chr22   18953122  chr22   22935373         inversion             5   \n",
       "9   chr22   20950573  chr22   20953841     deletion-like             4   \n",
       "10  chr22   23049707  chr22   23108069     deletion-like             2   \n",
       "11  chr22   23049910  chr22   23107858  duplication-like             2   \n",
       "\n",
       "   features orientation  homology_length homology_sequence  \n",
       "0   unknown          -+              5.0             GAGTG  \n",
       "1   unknown          ++              2.0                AT  \n",
       "2     ecDNA          --              3.0               AAC  \n",
       "3     ecDNA          +-              3.0               TGG  \n",
       "4     ecDNA          ++              6.0            ATAAAA  \n",
       "5     ecDNA          +-              NaN               NaN  \n",
       "6       NaN          ++              2.0                AA  \n",
       "7   unknown          +-              NaN               NaN  \n",
       "8   unknown          --              NaN               NaN  \n",
       "9   unknown          +-              NaN               NaN  \n",
       "10  unknown          +-              NaN               NaN  \n",
       "11    ecDNA          -+              NaN               NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>break_chrom1</th>\n",
       "      <th>break_pos1</th>\n",
       "      <th>break_chrom2</th>\n",
       "      <th>break_pos2</th>\n",
       "      <th>break_sv_type</th>\n",
       "      <th>break_read_support</th>\n",
       "      <th>break_features</th>\n",
       "      <th>break_orientation</th>\n",
       "      <th>AA_homology_len</th>\n",
       "      <th>AA_homology_seq</th>\n",
       "      <th>...</th>\n",
       "      <th>split</th>\n",
       "      <th>proper_pair</th>\n",
       "      <th>read_num</th>\n",
       "      <th>query_chrom</th>\n",
       "      <th>query_pos</th>\n",
       "      <th>query_end</th>\n",
       "      <th>query_orientation</th>\n",
       "      <th>query_cigar</th>\n",
       "      <th>query_aln_full</th>\n",
       "      <th>query_aln_sub</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607147</td>\n",
       "      <td>133607213</td>\n",
       "      <td>+</td>\n",
       "      <td>34S66M</td>\n",
       "      <td>TGACCACGGGACACCTTTGACCCTGGCCGCTGTGGAGTGGGTTTTA...</td>\n",
       "      <td>GAGTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632704</td>\n",
       "      <td>23632743</td>\n",
       "      <td>+</td>\n",
       "      <td>39M61H</td>\n",
       "      <td>TGACCACGGGACACCTTTGACCCTGGCCGCTGTGGAGTG</td>\n",
       "      <td>TGACCACGGGACACCTTTGACCCTGGCCGCTGTGGAGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607344</td>\n",
       "      <td>133607444</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>TGTTTCAAATTTTCAACTTTCAAAATTCCTGCTGCTTCTTTCCAAT...</td>\n",
       "      <td>TGTTTCAAATTTTCAACTTTCAAAATTCCTGCTGCTTCTTTCCAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632682</td>\n",
       "      <td>23632743</td>\n",
       "      <td>-</td>\n",
       "      <td>61M39S</td>\n",
       "      <td>GGAAACAGGGAGGTTGTTCAGATGACCACGGGACACCTTTGACCCT...</td>\n",
       "      <td>GGAAACAGGGAGGTTGTTCAGATGACCACGGGACACCTTTGACCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607147</td>\n",
       "      <td>133607191</td>\n",
       "      <td>-</td>\n",
       "      <td>56H44M</td>\n",
       "      <td>GAGTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTA</td>\n",
       "      <td>GAGTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632592</td>\n",
       "      <td>23632692</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>AGAGTTCAAGTAAGTACTGGTTTGGGGAGGAGGGTTGCAGCGGCCG...</td>\n",
       "      <td>AGAGTTCAAGTAAGTACTGGTTTGGGGAGGAGGGTTGCAGCGGCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632535</td>\n",
       "      <td>23632635</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>CTCCGGGGCTCTATGGGTTTCTGAATGTCATCGTCCACTCAGCCAC...</td>\n",
       "      <td>CTCCGGGGCTCTATGGGTTTCTGAATGTCATCGTCCACTCAGCCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607147</td>\n",
       "      <td>133607178</td>\n",
       "      <td>-</td>\n",
       "      <td>69H31M</td>\n",
       "      <td>GAGTGGGTTTTATCAGCTTCCATACCCAAAC</td>\n",
       "      <td>GAGTGGGTTTTATCAGCTTCCATACCCAAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632669</td>\n",
       "      <td>23632743</td>\n",
       "      <td>-</td>\n",
       "      <td>74M26S</td>\n",
       "      <td>GGGCAGGGTGTGGGGAAACAGGGAGGTTGTTCAGATGACCACGGGA...</td>\n",
       "      <td>GGGCAGGGTGTGGGGAAACAGGGAGGTTGTTCAGATGACCACGGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Discordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607147</td>\n",
       "      <td>133607218</td>\n",
       "      <td>-</td>\n",
       "      <td>29S71M</td>\n",
       "      <td>ACGGGACACCTTTGACCCTGGCCGCTGTGGAGTGGGTTTTATCAGC...</td>\n",
       "      <td>GAGTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Discordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632709</td>\n",
       "      <td>23632743</td>\n",
       "      <td>-</td>\n",
       "      <td>34M66H</td>\n",
       "      <td>ACGGGACACCTTTGACCCTGGCCGCTGTGGAGTG</td>\n",
       "      <td>ACGGGACACCTTTGACCCTGGCCGCTGTGGAGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Discordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632370</td>\n",
       "      <td>23632470</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>AGTATGTTTTTGGCCCATGACACTGGCTTACCTTGTGCCAGGCAGA...</td>\n",
       "      <td>AGTATGTTTTTGGCCCATGACACTGGCTTACCTTGTGCCAGGCAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Discordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607199</td>\n",
       "      <td>133607299</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>CATCTCTGATTGCACTAAATCTATAGGTTTCTTTGATCCTAGAAAT...</td>\n",
       "      <td>CATCTCTGATTGCACTAAATCTATAGGTTTCTTTGATCCTAGAAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Discordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607151</td>\n",
       "      <td>133607181</td>\n",
       "      <td>+</td>\n",
       "      <td>70H30M</td>\n",
       "      <td>GGGTTTTATCAGCTTCCATACCCAAACAGA</td>\n",
       "      <td>GGGTTTTATCAGCTTCCATACCCAAACAGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Discordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632672</td>\n",
       "      <td>23632741</td>\n",
       "      <td>+</td>\n",
       "      <td>69M31S</td>\n",
       "      <td>CAGGGTGTGGGGAAACAGGGAGGTTGTTCAGATGACCACGGGACAC...</td>\n",
       "      <td>CAGGGTGTGGGGAAACAGGGAGGTTGTTCAGATGACCACGGGACAC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   break_chrom1  break_pos1 break_chrom2  break_pos2     break_sv_type  \\\n",
       "0          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "1          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "2          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "3          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "4          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "5          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "6          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "7          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "8          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "9          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "10         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "11         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "12         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "13         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "14         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "\n",
       "    break_read_support break_features break_orientation  AA_homology_len  \\\n",
       "0                    7        unknown                -+              5.0   \n",
       "1                    7        unknown                -+              5.0   \n",
       "2                    7        unknown                -+              5.0   \n",
       "3                    7        unknown                -+              5.0   \n",
       "4                    7        unknown                -+              5.0   \n",
       "5                    7        unknown                -+              5.0   \n",
       "6                    7        unknown                -+              5.0   \n",
       "7                    7        unknown                -+              5.0   \n",
       "8                    7        unknown                -+              5.0   \n",
       "9                    7        unknown                -+              5.0   \n",
       "10                   7        unknown                -+              5.0   \n",
       "11                   7        unknown                -+              5.0   \n",
       "12                   7        unknown                -+              5.0   \n",
       "13                   7        unknown                -+              5.0   \n",
       "14                   7        unknown                -+              5.0   \n",
       "\n",
       "   AA_homology_seq  ...  split proper_pair  read_num query_chrom  query_pos  \\\n",
       "0            GAGTG  ...   True  Concordant         1        chr9  133607147   \n",
       "1            GAGTG  ...   True  Concordant         1       chr22   23632704   \n",
       "2            GAGTG  ...  False  Concordant         2        chr9  133607344   \n",
       "3            GAGTG  ...   True  Concordant         1       chr22   23632682   \n",
       "4            GAGTG  ...   True  Concordant         1        chr9  133607147   \n",
       "5            GAGTG  ...  False  Concordant         2       chr22   23632592   \n",
       "6            GAGTG  ...  False  Concordant         1       chr22   23632535   \n",
       "7            GAGTG  ...   True  Concordant         2        chr9  133607147   \n",
       "8            GAGTG  ...   True  Concordant         2       chr22   23632669   \n",
       "9            GAGTG  ...   True  Discordant         1        chr9  133607147   \n",
       "10           GAGTG  ...   True  Discordant         1       chr22   23632709   \n",
       "11           GAGTG  ...  False  Discordant         2       chr22   23632370   \n",
       "12           GAGTG  ...  False  Discordant         1        chr9  133607199   \n",
       "13           GAGTG  ...   True  Discordant         2        chr9  133607151   \n",
       "14           GAGTG  ...   True  Discordant         2       chr22   23632672   \n",
       "\n",
       "    query_end  query_orientation  query_cigar  \\\n",
       "0   133607213                  +       34S66M   \n",
       "1    23632743                  +       39M61H   \n",
       "2   133607444                  -         100M   \n",
       "3    23632743                  -       61M39S   \n",
       "4   133607191                  -       56H44M   \n",
       "5    23632692                  +         100M   \n",
       "6    23632635                  +         100M   \n",
       "7   133607178                  -       69H31M   \n",
       "8    23632743                  -       74M26S   \n",
       "9   133607218                  -       29S71M   \n",
       "10   23632743                  -       34M66H   \n",
       "11   23632470                  +         100M   \n",
       "12  133607299                  -         100M   \n",
       "13  133607181                  +       70H30M   \n",
       "14   23632741                  +       69M31S   \n",
       "\n",
       "                                       query_aln_full  \\\n",
       "0   TGACCACGGGACACCTTTGACCCTGGCCGCTGTGGAGTGGGTTTTA...   \n",
       "1             TGACCACGGGACACCTTTGACCCTGGCCGCTGTGGAGTG   \n",
       "2   TGTTTCAAATTTTCAACTTTCAAAATTCCTGCTGCTTCTTTCCAAT...   \n",
       "3   GGAAACAGGGAGGTTGTTCAGATGACCACGGGACACCTTTGACCCT...   \n",
       "4        GAGTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTA   \n",
       "5   AGAGTTCAAGTAAGTACTGGTTTGGGGAGGAGGGTTGCAGCGGCCG...   \n",
       "6   CTCCGGGGCTCTATGGGTTTCTGAATGTCATCGTCCACTCAGCCAC...   \n",
       "7                     GAGTGGGTTTTATCAGCTTCCATACCCAAAC   \n",
       "8   GGGCAGGGTGTGGGGAAACAGGGAGGTTGTTCAGATGACCACGGGA...   \n",
       "9   ACGGGACACCTTTGACCCTGGCCGCTGTGGAGTGGGTTTTATCAGC...   \n",
       "10                 ACGGGACACCTTTGACCCTGGCCGCTGTGGAGTG   \n",
       "11  AGTATGTTTTTGGCCCATGACACTGGCTTACCTTGTGCCAGGCAGA...   \n",
       "12  CATCTCTGATTGCACTAAATCTATAGGTTTCTTTGATCCTAGAAAT...   \n",
       "13                     GGGTTTTATCAGCTTCCATACCCAAACAGA   \n",
       "14  CAGGGTGTGGGGAAACAGGGAGGTTGTTCAGATGACCACGGGACAC...   \n",
       "\n",
       "                                        query_aln_sub  \n",
       "0   GAGTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAG...  \n",
       "1             TGACCACGGGACACCTTTGACCCTGGCCGCTGTGGAGTG  \n",
       "2   TGTTTCAAATTTTCAACTTTCAAAATTCCTGCTGCTTCTTTCCAAT...  \n",
       "3   GGAAACAGGGAGGTTGTTCAGATGACCACGGGACACCTTTGACCCT...  \n",
       "4        GAGTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTA  \n",
       "5   AGAGTTCAAGTAAGTACTGGTTTGGGGAGGAGGGTTGCAGCGGCCG...  \n",
       "6   CTCCGGGGCTCTATGGGTTTCTGAATGTCATCGTCCACTCAGCCAC...  \n",
       "7                     GAGTGGGTTTTATCAGCTTCCATACCCAAAC  \n",
       "8   GGGCAGGGTGTGGGGAAACAGGGAGGTTGTTCAGATGACCACGGGA...  \n",
       "9   GAGTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAG...  \n",
       "10                 ACGGGACACCTTTGACCCTGGCCGCTGTGGAGTG  \n",
       "11  AGTATGTTTTTGGCCCATGACACTGGCTTACCTTGTGCCAGGCAGA...  \n",
       "12  CATCTCTGATTGCACTAAATCTATAGGTTTCTTTGATCCTAGAAAT...  \n",
       "13                     GGGTTTTATCAGCTTCCATACCCAAACAGA  \n",
       "14  CAGGGTGTGGGGAAACAGGGAGGTTGTTCAGATGACCACGGGACAC...  \n",
       "\n",
       "[15 rows x 22 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>break_chrom1</th>\n",
       "      <th>break_pos1</th>\n",
       "      <th>break_chrom2</th>\n",
       "      <th>break_pos2</th>\n",
       "      <th>break_sv_type</th>\n",
       "      <th>break_read_support</th>\n",
       "      <th>break_features</th>\n",
       "      <th>break_orientation</th>\n",
       "      <th>AA_homology_len</th>\n",
       "      <th>AA_homology_seq</th>\n",
       "      <th>...</th>\n",
       "      <th>split</th>\n",
       "      <th>proper_pair</th>\n",
       "      <th>read_num</th>\n",
       "      <th>query_chrom</th>\n",
       "      <th>query_pos</th>\n",
       "      <th>query_end</th>\n",
       "      <th>query_orientation</th>\n",
       "      <th>query_cigar</th>\n",
       "      <th>query_aln_full</th>\n",
       "      <th>query_aln_sub</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133606903</td>\n",
       "      <td>133607003</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>CCCTCATATTGACCTAGCTGTTTAATTGAAAAGAACATCTGCTGCT...</td>\n",
       "      <td>CCCTCATATTGACCTAGCTGTTTAATTGAAAAGAACATCTGCTGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607120</td>\n",
       "      <td>133607220</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>CTCCTTTGTATTTCCATATACATTTTAGAGTGGGTTTTATCAGCTT...</td>\n",
       "      <td>CTCCTTTGTATTTCCATATACATTTTAGAGTGGGTTTTATCAGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133606645</td>\n",
       "      <td>133606745</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>TGGCCTCCCAAAGTGCTGGGGTTACAGACGTGAGCCACTGCGTCCG...</td>\n",
       "      <td>TGGCCTCCCAAAGTGCTGGGGTTACAGACGTGAGCCACTGCGTCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607005</td>\n",
       "      <td>133607105</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>TTTGTTTGTCTTTCCTTGCACCAGTATTACAATGTATTACTGTAGC...</td>\n",
       "      <td>TTTGTTTGTCTTTCCTTGCACCAGTATTACAATGTATTACTGTAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607067</td>\n",
       "      <td>133607167</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>ATGCCTGGTAGAATAAGGCCTCCTGTTTTGTTCTTCAATATTGTCT...</td>\n",
       "      <td>ATGCCTGGTAGAATAAGGCCTCCTGTTTTGTTCTTCAATATTGTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133606727</td>\n",
       "      <td>133606827</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>ACCTTAATTTTGACTTAGTCTAATTTAATCTTTTATGGTTAGTACT...</td>\n",
       "      <td>ACCTTAATTTTGACTTAGTCTAATTTAATCTTTTATGGTTAGTACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607147</td>\n",
       "      <td>133607226</td>\n",
       "      <td>+</td>\n",
       "      <td>21S79M</td>\n",
       "      <td>CCTTTGACCCTGGCCGCTGTGGAGTGGGTTTTATCAGCTTCCATAC...</td>\n",
       "      <td>GAGTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607451</td>\n",
       "      <td>133607551</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>CCCTTGATAGTGGAGACCATGTTTATACATTTTCATGGATCGAACA...</td>\n",
       "      <td>CCCTTGATAGTGGAGACCATGTTTATACATTTTCATGGATCGAACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607186</td>\n",
       "      <td>133607286</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>CCTTAAGGATTTTCTTCTCTGATTGCACTAAATCTATAGGTTTCTT...</td>\n",
       "      <td>CCTTAAGGATTTTCTTCTCTGATTGCACTAAATCTATAGGTTTCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607147</td>\n",
       "      <td>133607224</td>\n",
       "      <td>+</td>\n",
       "      <td>23S77M</td>\n",
       "      <td>CACCTTTGACCCTGGCCGCTGTGGAGTGGGTTTTATCAGCTTCCAT...</td>\n",
       "      <td>GAGTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Discordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607147</td>\n",
       "      <td>133607241</td>\n",
       "      <td>-</td>\n",
       "      <td>6S94M</td>\n",
       "      <td>GCTGTGGAGTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACC...</td>\n",
       "      <td>GAGTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Discordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632298</td>\n",
       "      <td>23632398</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>CTTCGAGTCACTGGTTTGCCTGTATTGTGAAACCAACTGGATCCTG...</td>\n",
       "      <td>CTTCGAGTCACTGGTTTGCCTGTATTGTGAAACCAACTGGATCCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Discordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607149</td>\n",
       "      <td>133607249</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>GTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAGGA...</td>\n",
       "      <td>GTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Discordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632385</td>\n",
       "      <td>23632485</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>CATGACACTGGCTTACCTTGTGCCAGGCAGATGGCAGCCACACAGT...</td>\n",
       "      <td>CATGACACTGGCTTACCTTGTGCCAGGCAGATGGCAGCCACACAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607153</td>\n",
       "      <td>133607253</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>GTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAGGATTTT...</td>\n",
       "      <td>GTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAGGATTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607562</td>\n",
       "      <td>133607662</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>CAGGCATTGAACAAGTACTAAGTGCCTGTTCTGTATCAGGTACTGT...</td>\n",
       "      <td>CAGGCATTGAACAAGTACTAAGTGCCTGTTCTGTATCAGGTACTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Discordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632554</td>\n",
       "      <td>23632654</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>TCTGAATGTCATCGTCCACTCAGCCACTGGATTTAAGCAGAGTTCA...</td>\n",
       "      <td>TCTGAATGTCATCGTCCACTCAGCCACTGGATTTAAGCAGAGTTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Discordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607181</td>\n",
       "      <td>133607281</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>AATACCCTTAAGGATTTTCTTCTCTGATTGCACTAAATCTATAGGT...</td>\n",
       "      <td>AATACCCTTAAGGATTTTCTTCTCTGATTGCACTAAATCTATAGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607269</td>\n",
       "      <td>133607369</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>TATTGATGAAAGAAAGCTATTTCAGGAAACTTTTTTTTTTCTGTTA...</td>\n",
       "      <td>TATTGATGAAAGAAAGCTATTTCAGGAAACTTTTTTTTTTCTGTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607189</td>\n",
       "      <td>133607289</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>TAAGGATTTTCTTCTCTGATTGCACTAAATCTATAGGTTTCTTTGA...</td>\n",
       "      <td>TAAGGATTTTCTTCTCTGATTGCACTAAATCTATAGGTTTCTTTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607271</td>\n",
       "      <td>133607371</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>TTGATGAAAGAAAGCTATTTCAGGAAACTTTTTTTTTTCTGTTAAT...</td>\n",
       "      <td>TTGATGAAAGAAAGCTATTTCAGGAAACTTTTTTTTTTCTGTTAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607195</td>\n",
       "      <td>133607295</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>TTTTCTTCTCTGATTGCACTAAATCTATAGGTTTCTTTGATCCTAG...</td>\n",
       "      <td>TTTTCTTCTCTGATTGCACTAAATCTATAGGTTTCTTTGATCCTAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607356</td>\n",
       "      <td>133607457</td>\n",
       "      <td>-</td>\n",
       "      <td>50M1D50M</td>\n",
       "      <td>TCAACTTTCAAAATACCTGCTGCTTCTTTCCAATTCACCTTTCTCT...</td>\n",
       "      <td>TCAACTTTCAAAATACCTGCTGCTTCTTTCCAATTCACCTTTCTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607201</td>\n",
       "      <td>133607301</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>TCTCTGATTGCACTAAATCTATAGGTTTCTTTGATCCTAGAAATGA...</td>\n",
       "      <td>TCTCTGATTGCACTAAATCTATAGGTTTCTTTGATCCTAGAAATGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Discordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607203</td>\n",
       "      <td>133607304</td>\n",
       "      <td>-</td>\n",
       "      <td>56M1D44M</td>\n",
       "      <td>TCTGATTGCACTAAATCTATAGGTTTCTTTGATCCTAGAAATGAAT...</td>\n",
       "      <td>TCTGATTGCACTAAATCTATAGGTTTCTTTGATCCTAGAAATGAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Discordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632475</td>\n",
       "      <td>23632575</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>GCCTCCCTTTCCCGGGACAACAGAAGCTGACCTCTTTGATCTCTTG...</td>\n",
       "      <td>GCCTCCCTTTCCCGGGACAACAGAAGCTGACCTCTTTGATCTCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607635</td>\n",
       "      <td>133607735</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>CAACATAGCCAACAAAAGGCTCTTTGTGTCTAATTTATATTCCTGA...</td>\n",
       "      <td>CAACATAGCCAACAAAAGGCTCTTTGTGTCTAATTTATATTCCTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607234</td>\n",
       "      <td>133607334</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>ATCCTAGAAATGAATAATAGAATATAAAGAATTCTTATTGATGAAA...</td>\n",
       "      <td>ATCCTAGAAATGAATAATAGAATATAAAGAATTCTTATTGATGAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Discordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133607251</td>\n",
       "      <td>133607351</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>TAGAATATAAAGAATTCTTATTGATGAAAGAAAGCTATTTCAGGAA...</td>\n",
       "      <td>TAGAATATAAAGAATTCTTATTGATGAAAGAAAGCTATTTCAGGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Discordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632486</td>\n",
       "      <td>23632586</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>CCGGGACAACAGAAGCTGACCTCTTTGATCTCTTGCGCAGATGATG...</td>\n",
       "      <td>CCGGGACAACAGAAGCTGACCTCTTTGATCTCTTGCGCAGATGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632073</td>\n",
       "      <td>23632173</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>CCTCAGATGCTCTGTGCCTTGGATCTGGCCCCACTCCCGTCCTCCC...</td>\n",
       "      <td>CCTCAGATGCTCTGTGCCTTGGATCTGGCCCCACTCCCGTCCTCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632532</td>\n",
       "      <td>23632632</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>AGTCTCCGGGGCTCTATGGGTTTCTGAATGTCATCGTCCACTCAGC...</td>\n",
       "      <td>AGTCTCCGGGGCTCTATGGGTTTCTGAATGTCATCGTCCACTCAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632609</td>\n",
       "      <td>23632709</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>TGGTTTGGGGAGGAGGGTTGCAGCGGCCGAGCCAGGGTCTCCACCC...</td>\n",
       "      <td>TGGTTTGGGGAGGAGGGTTGCAGCGGCCGAGCCAGGGTCTCCACCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632538</td>\n",
       "      <td>23632638</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>CGGGGCTCTATGGGTTTCTGAATGTCATCGTCCACTCAGCCACTGG...</td>\n",
       "      <td>CGGGGCTCTATGGGTTTCTGAATGTCATCGTCCACTCAGCCACTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632553</td>\n",
       "      <td>23632653</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>TTCTGAATGTCATCGTCCACTCAGCCACTGGATTTAAGCAGAGTTC...</td>\n",
       "      <td>TTCTGAATGTCATCGTCCACTCAGCCACTGGATTTAAGCAGAGTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632376</td>\n",
       "      <td>23632476</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>TTTTTGGCCCATGACACTGGCTTACCTTGTGCCAGGCAGATGGCAG...</td>\n",
       "      <td>TTTTTGGCCCATGACACTGGCTTACCTTGTGCCAGGCAGATGGCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632572</td>\n",
       "      <td>23632672</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>CTCAGACACTGGATTTAAGCAGATTTCAAGTAAGTACTGGTTTGGG...</td>\n",
       "      <td>CTCAGACACTGGATTTAAGCAGATTTCAAGTAAGTACTGGTTTGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632371</td>\n",
       "      <td>23632459</td>\n",
       "      <td>+</td>\n",
       "      <td>12S88M</td>\n",
       "      <td>AAATAAAGCTACGTATGTTTTTAGCCAAAGACACTCGCTTACCTTG...</td>\n",
       "      <td>GTATGTTTTTAGCCAAAGACACTCGCTTACCTTGTGCCAGGCAGAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632578</td>\n",
       "      <td>23632678</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>CACTGGATTTAAGCAGAGTTCAAGTAAGTACTGGTTTGGGGAGGAG...</td>\n",
       "      <td>CACTGGATTTAAGCAGAGTTCAAGTAAGTACTGGTTTGGGGAGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632288</td>\n",
       "      <td>23632388</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>CCTTATTACACTTCGAGTCACTGGTTTGCCTGTATTGTGAAACCAA...</td>\n",
       "      <td>CCTTATTACACTTCGAGTCACTGGTTTGCCTGTATTGTGAAACCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632606</td>\n",
       "      <td>23632706</td>\n",
       "      <td>-</td>\n",
       "      <td>100M</td>\n",
       "      <td>TACTGGTTTGGGGAGGAGGGTTGCAGCGGCCGAGCCAGGGTCTCCA...</td>\n",
       "      <td>TACTGGTTTGGGGAGGAGGGTTGCAGCGGCCGAGCCAGGGTCTCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632476</td>\n",
       "      <td>23632576</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>CCTCCCTTTCCCGGGACAACAGAAGCTGACCTCTTTGATCTCTTGC...</td>\n",
       "      <td>CCTCCCTTTCCCGGGACAACAGAAGCTGACCTCTTTGATCTCTTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632306</td>\n",
       "      <td>23632406</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>CACTGGTTTGCCTGTATTGTGAAACCAACTGGATCCTGAGATCCCC...</td>\n",
       "      <td>CACTGGTTTGCCTGTATTGTGAAACCAACTGGATCCTGAGATCCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632647</td>\n",
       "      <td>23632743</td>\n",
       "      <td>-</td>\n",
       "      <td>96M4S</td>\n",
       "      <td>CTCCACCCAGGAAGGACTAATCGGGCAGGGTGTGGGGAAACAGGGA...</td>\n",
       "      <td>CTCCACCCAGGAAGGACTAATCGGGCAGGGTGTGGGGAAACAGGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632269</td>\n",
       "      <td>23632369</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>CCCTCTGCTGTCCTTGGAACCTTATTACACTTCGAGTCACTGGTTT...</td>\n",
       "      <td>CCCTCTGCTGTCCTTGGAACCTTATTACACTTCGAGTCACTGGTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632654</td>\n",
       "      <td>23632743</td>\n",
       "      <td>-</td>\n",
       "      <td>89M11S</td>\n",
       "      <td>CAGGAAGGACTAATCGGGCAGGGTGTGGGGAAACAGGGAGGTTGTT...</td>\n",
       "      <td>CAGGAAGGACTAATCGGGCAGGGTGTGGGGAAACAGGGAGGTTGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632347</td>\n",
       "      <td>23632447</td>\n",
       "      <td>+</td>\n",
       "      <td>100M</td>\n",
       "      <td>TCCCCAAGACAGAAATCATGATGAGTATGTTTTTGGCCCATGACAC...</td>\n",
       "      <td>TCCCCAAGACAGAAATCATGATGAGTATGTTTTTGGCCCATGACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133607146</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632741</td>\n",
       "      <td>interchromosomal</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-+</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Concordant</td>\n",
       "      <td>2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23632655</td>\n",
       "      <td>23632743</td>\n",
       "      <td>-</td>\n",
       "      <td>88M12S</td>\n",
       "      <td>AGGAAGGACTAATCGGGCAGGGTGTGGGGAAACAGGGAGGTTGTTC...</td>\n",
       "      <td>AGGAAGGACTAATCGGGCAGGGTGTGGGGAAACAGGGAGGTTGTTC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>48 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   break_chrom1  break_pos1 break_chrom2  break_pos2     break_sv_type  \\\n",
       "0          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "1          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "2          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "3          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "4          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "5          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "6          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "7          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "8          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "9          chr9   133607146        chr22    23632741  interchromosomal   \n",
       "10         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "11         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "12         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "13         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "14         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "15         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "16         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "17         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "18         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "19         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "20         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "21         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "22         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "23         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "24         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "25         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "26         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "27         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "28         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "29         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "30         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "31         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "32         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "33         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "34         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "35         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "36         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "37         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "38         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "39         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "40         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "41         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "42         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "43         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "44         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "45         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "46         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "47         chr9   133607146        chr22    23632741  interchromosomal   \n",
       "\n",
       "    break_read_support break_features break_orientation  AA_homology_len  \\\n",
       "0                    7        unknown                -+              5.0   \n",
       "1                    7        unknown                -+              5.0   \n",
       "2                    7        unknown                -+              5.0   \n",
       "3                    7        unknown                -+              5.0   \n",
       "4                    7        unknown                -+              5.0   \n",
       "5                    7        unknown                -+              5.0   \n",
       "6                    7        unknown                -+              5.0   \n",
       "7                    7        unknown                -+              5.0   \n",
       "8                    7        unknown                -+              5.0   \n",
       "9                    7        unknown                -+              5.0   \n",
       "10                   7        unknown                -+              5.0   \n",
       "11                   7        unknown                -+              5.0   \n",
       "12                   7        unknown                -+              5.0   \n",
       "13                   7        unknown                -+              5.0   \n",
       "14                   7        unknown                -+              5.0   \n",
       "15                   7        unknown                -+              5.0   \n",
       "16                   7        unknown                -+              5.0   \n",
       "17                   7        unknown                -+              5.0   \n",
       "18                   7        unknown                -+              5.0   \n",
       "19                   7        unknown                -+              5.0   \n",
       "20                   7        unknown                -+              5.0   \n",
       "21                   7        unknown                -+              5.0   \n",
       "22                   7        unknown                -+              5.0   \n",
       "23                   7        unknown                -+              5.0   \n",
       "24                   7        unknown                -+              5.0   \n",
       "25                   7        unknown                -+              5.0   \n",
       "26                   7        unknown                -+              5.0   \n",
       "27                   7        unknown                -+              5.0   \n",
       "28                   7        unknown                -+              5.0   \n",
       "29                   7        unknown                -+              5.0   \n",
       "30                   7        unknown                -+              5.0   \n",
       "31                   7        unknown                -+              5.0   \n",
       "32                   7        unknown                -+              5.0   \n",
       "33                   7        unknown                -+              5.0   \n",
       "34                   7        unknown                -+              5.0   \n",
       "35                   7        unknown                -+              5.0   \n",
       "36                   7        unknown                -+              5.0   \n",
       "37                   7        unknown                -+              5.0   \n",
       "38                   7        unknown                -+              5.0   \n",
       "39                   7        unknown                -+              5.0   \n",
       "40                   7        unknown                -+              5.0   \n",
       "41                   7        unknown                -+              5.0   \n",
       "42                   7        unknown                -+              5.0   \n",
       "43                   7        unknown                -+              5.0   \n",
       "44                   7        unknown                -+              5.0   \n",
       "45                   7        unknown                -+              5.0   \n",
       "46                   7        unknown                -+              5.0   \n",
       "47                   7        unknown                -+              5.0   \n",
       "\n",
       "   AA_homology_seq  ...  split proper_pair  read_num query_chrom  query_pos  \\\n",
       "0            GAGTG  ...  False  Concordant         1        chr9  133606903   \n",
       "1            GAGTG  ...  False  Concordant         2        chr9  133607120   \n",
       "2            GAGTG  ...  False  Concordant         1        chr9  133606645   \n",
       "3            GAGTG  ...  False  Concordant         2        chr9  133607005   \n",
       "4            GAGTG  ...  False  Concordant         1        chr9  133607067   \n",
       "5            GAGTG  ...  False  Concordant         2        chr9  133606727   \n",
       "6            GAGTG  ...  False  Concordant         1        chr9  133607147   \n",
       "7            GAGTG  ...  False  Concordant         2        chr9  133607451   \n",
       "8            GAGTG  ...  False  Concordant         1        chr9  133607186   \n",
       "9            GAGTG  ...  False  Concordant         2        chr9  133607147   \n",
       "10           GAGTG  ...  False  Discordant         1        chr9  133607147   \n",
       "11           GAGTG  ...  False  Discordant         2       chr22   23632298   \n",
       "12           GAGTG  ...  False  Discordant         1        chr9  133607149   \n",
       "13           GAGTG  ...  False  Discordant         2       chr22   23632385   \n",
       "14           GAGTG  ...  False  Concordant         1        chr9  133607153   \n",
       "15           GAGTG  ...  False  Concordant         2        chr9  133607562   \n",
       "16           GAGTG  ...  False  Discordant         1       chr22   23632554   \n",
       "17           GAGTG  ...  False  Discordant         2        chr9  133607181   \n",
       "18           GAGTG  ...  False  Concordant         1        chr9  133607269   \n",
       "19           GAGTG  ...  False  Concordant         2        chr9  133607189   \n",
       "20           GAGTG  ...  False  Concordant         1        chr9  133607271   \n",
       "21           GAGTG  ...  False  Concordant         2        chr9  133607195   \n",
       "22           GAGTG  ...  False  Concordant         1        chr9  133607356   \n",
       "23           GAGTG  ...  False  Concordant         2        chr9  133607201   \n",
       "24           GAGTG  ...  False  Discordant         1        chr9  133607203   \n",
       "25           GAGTG  ...  False  Discordant         2       chr22   23632475   \n",
       "26           GAGTG  ...  False  Concordant         1        chr9  133607635   \n",
       "27           GAGTG  ...  False  Concordant         2        chr9  133607234   \n",
       "28           GAGTG  ...  False  Discordant         1        chr9  133607251   \n",
       "29           GAGTG  ...  False  Discordant         2       chr22   23632486   \n",
       "30           GAGTG  ...  False  Concordant         1       chr22   23632073   \n",
       "31           GAGTG  ...  False  Concordant         2       chr22   23632532   \n",
       "32           GAGTG  ...  False  Concordant         1       chr22   23632609   \n",
       "33           GAGTG  ...  False  Concordant         2       chr22   23632538   \n",
       "34           GAGTG  ...  False  Concordant         1       chr22   23632553   \n",
       "35           GAGTG  ...  False  Concordant         2       chr22   23632376   \n",
       "36           GAGTG  ...  False  Concordant         1       chr22   23632572   \n",
       "37           GAGTG  ...  False  Concordant         2       chr22   23632371   \n",
       "38           GAGTG  ...  False  Concordant         1       chr22   23632578   \n",
       "39           GAGTG  ...  False  Concordant         2       chr22   23632288   \n",
       "40           GAGTG  ...  False  Concordant         1       chr22   23632606   \n",
       "41           GAGTG  ...  False  Concordant         2       chr22   23632476   \n",
       "42           GAGTG  ...  False  Concordant         1       chr22   23632306   \n",
       "43           GAGTG  ...  False  Concordant         2       chr22   23632647   \n",
       "44           GAGTG  ...  False  Concordant         1       chr22   23632269   \n",
       "45           GAGTG  ...  False  Concordant         2       chr22   23632654   \n",
       "46           GAGTG  ...  False  Concordant         1       chr22   23632347   \n",
       "47           GAGTG  ...  False  Concordant         2       chr22   23632655   \n",
       "\n",
       "    query_end  query_orientation  query_cigar  \\\n",
       "0   133607003                  +         100M   \n",
       "1   133607220                  -         100M   \n",
       "2   133606745                  +         100M   \n",
       "3   133607105                  -         100M   \n",
       "4   133607167                  -         100M   \n",
       "5   133606827                  +         100M   \n",
       "6   133607226                  +       21S79M   \n",
       "7   133607551                  -         100M   \n",
       "8   133607286                  -         100M   \n",
       "9   133607224                  +       23S77M   \n",
       "10  133607241                  -        6S94M   \n",
       "11   23632398                  +         100M   \n",
       "12  133607249                  -         100M   \n",
       "13   23632485                  +         100M   \n",
       "14  133607253                  +         100M   \n",
       "15  133607662                  -         100M   \n",
       "16   23632654                  +         100M   \n",
       "17  133607281                  -         100M   \n",
       "18  133607369                  -         100M   \n",
       "19  133607289                  +         100M   \n",
       "20  133607371                  -         100M   \n",
       "21  133607295                  +         100M   \n",
       "22  133607457                  -     50M1D50M   \n",
       "23  133607301                  +         100M   \n",
       "24  133607304                  -     56M1D44M   \n",
       "25   23632575                  +         100M   \n",
       "26  133607735                  -         100M   \n",
       "27  133607334                  +         100M   \n",
       "28  133607351                  -         100M   \n",
       "29   23632586                  +         100M   \n",
       "30   23632173                  +         100M   \n",
       "31   23632632                  -         100M   \n",
       "32   23632709                  -         100M   \n",
       "33   23632638                  +         100M   \n",
       "34   23632653                  -         100M   \n",
       "35   23632476                  +         100M   \n",
       "36   23632672                  -         100M   \n",
       "37   23632459                  +       12S88M   \n",
       "38   23632678                  -         100M   \n",
       "39   23632388                  +         100M   \n",
       "40   23632706                  -         100M   \n",
       "41   23632576                  +         100M   \n",
       "42   23632406                  +         100M   \n",
       "43   23632743                  -        96M4S   \n",
       "44   23632369                  +         100M   \n",
       "45   23632743                  -       89M11S   \n",
       "46   23632447                  +         100M   \n",
       "47   23632743                  -       88M12S   \n",
       "\n",
       "                                       query_aln_full  \\\n",
       "0   CCCTCATATTGACCTAGCTGTTTAATTGAAAAGAACATCTGCTGCT...   \n",
       "1   CTCCTTTGTATTTCCATATACATTTTAGAGTGGGTTTTATCAGCTT...   \n",
       "2   TGGCCTCCCAAAGTGCTGGGGTTACAGACGTGAGCCACTGCGTCCG...   \n",
       "3   TTTGTTTGTCTTTCCTTGCACCAGTATTACAATGTATTACTGTAGC...   \n",
       "4   ATGCCTGGTAGAATAAGGCCTCCTGTTTTGTTCTTCAATATTGTCT...   \n",
       "5   ACCTTAATTTTGACTTAGTCTAATTTAATCTTTTATGGTTAGTACT...   \n",
       "6   CCTTTGACCCTGGCCGCTGTGGAGTGGGTTTTATCAGCTTCCATAC...   \n",
       "7   CCCTTGATAGTGGAGACCATGTTTATACATTTTCATGGATCGAACA...   \n",
       "8   CCTTAAGGATTTTCTTCTCTGATTGCACTAAATCTATAGGTTTCTT...   \n",
       "9   CACCTTTGACCCTGGCCGCTGTGGAGTGGGTTTTATCAGCTTCCAT...   \n",
       "10  GCTGTGGAGTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACC...   \n",
       "11  CTTCGAGTCACTGGTTTGCCTGTATTGTGAAACCAACTGGATCCTG...   \n",
       "12  GTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAGGA...   \n",
       "13  CATGACACTGGCTTACCTTGTGCCAGGCAGATGGCAGCCACACAGT...   \n",
       "14  GTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAGGATTTT...   \n",
       "15  CAGGCATTGAACAAGTACTAAGTGCCTGTTCTGTATCAGGTACTGT...   \n",
       "16  TCTGAATGTCATCGTCCACTCAGCCACTGGATTTAAGCAGAGTTCA...   \n",
       "17  AATACCCTTAAGGATTTTCTTCTCTGATTGCACTAAATCTATAGGT...   \n",
       "18  TATTGATGAAAGAAAGCTATTTCAGGAAACTTTTTTTTTTCTGTTA...   \n",
       "19  TAAGGATTTTCTTCTCTGATTGCACTAAATCTATAGGTTTCTTTGA...   \n",
       "20  TTGATGAAAGAAAGCTATTTCAGGAAACTTTTTTTTTTCTGTTAAT...   \n",
       "21  TTTTCTTCTCTGATTGCACTAAATCTATAGGTTTCTTTGATCCTAG...   \n",
       "22  TCAACTTTCAAAATACCTGCTGCTTCTTTCCAATTCACCTTTCTCT...   \n",
       "23  TCTCTGATTGCACTAAATCTATAGGTTTCTTTGATCCTAGAAATGA...   \n",
       "24  TCTGATTGCACTAAATCTATAGGTTTCTTTGATCCTAGAAATGAAT...   \n",
       "25  GCCTCCCTTTCCCGGGACAACAGAAGCTGACCTCTTTGATCTCTTG...   \n",
       "26  CAACATAGCCAACAAAAGGCTCTTTGTGTCTAATTTATATTCCTGA...   \n",
       "27  ATCCTAGAAATGAATAATAGAATATAAAGAATTCTTATTGATGAAA...   \n",
       "28  TAGAATATAAAGAATTCTTATTGATGAAAGAAAGCTATTTCAGGAA...   \n",
       "29  CCGGGACAACAGAAGCTGACCTCTTTGATCTCTTGCGCAGATGATG...   \n",
       "30  CCTCAGATGCTCTGTGCCTTGGATCTGGCCCCACTCCCGTCCTCCC...   \n",
       "31  AGTCTCCGGGGCTCTATGGGTTTCTGAATGTCATCGTCCACTCAGC...   \n",
       "32  TGGTTTGGGGAGGAGGGTTGCAGCGGCCGAGCCAGGGTCTCCACCC...   \n",
       "33  CGGGGCTCTATGGGTTTCTGAATGTCATCGTCCACTCAGCCACTGG...   \n",
       "34  TTCTGAATGTCATCGTCCACTCAGCCACTGGATTTAAGCAGAGTTC...   \n",
       "35  TTTTTGGCCCATGACACTGGCTTACCTTGTGCCAGGCAGATGGCAG...   \n",
       "36  CTCAGACACTGGATTTAAGCAGATTTCAAGTAAGTACTGGTTTGGG...   \n",
       "37  AAATAAAGCTACGTATGTTTTTAGCCAAAGACACTCGCTTACCTTG...   \n",
       "38  CACTGGATTTAAGCAGAGTTCAAGTAAGTACTGGTTTGGGGAGGAG...   \n",
       "39  CCTTATTACACTTCGAGTCACTGGTTTGCCTGTATTGTGAAACCAA...   \n",
       "40  TACTGGTTTGGGGAGGAGGGTTGCAGCGGCCGAGCCAGGGTCTCCA...   \n",
       "41  CCTCCCTTTCCCGGGACAACAGAAGCTGACCTCTTTGATCTCTTGC...   \n",
       "42  CACTGGTTTGCCTGTATTGTGAAACCAACTGGATCCTGAGATCCCC...   \n",
       "43  CTCCACCCAGGAAGGACTAATCGGGCAGGGTGTGGGGAAACAGGGA...   \n",
       "44  CCCTCTGCTGTCCTTGGAACCTTATTACACTTCGAGTCACTGGTTT...   \n",
       "45  CAGGAAGGACTAATCGGGCAGGGTGTGGGGAAACAGGGAGGTTGTT...   \n",
       "46  TCCCCAAGACAGAAATCATGATGAGTATGTTTTTGGCCCATGACAC...   \n",
       "47  AGGAAGGACTAATCGGGCAGGGTGTGGGGAAACAGGGAGGTTGTTC...   \n",
       "\n",
       "                                        query_aln_sub  \n",
       "0   CCCTCATATTGACCTAGCTGTTTAATTGAAAAGAACATCTGCTGCT...  \n",
       "1   CTCCTTTGTATTTCCATATACATTTTAGAGTGGGTTTTATCAGCTT...  \n",
       "2   TGGCCTCCCAAAGTGCTGGGGTTACAGACGTGAGCCACTGCGTCCG...  \n",
       "3   TTTGTTTGTCTTTCCTTGCACCAGTATTACAATGTATTACTGTAGC...  \n",
       "4   ATGCCTGGTAGAATAAGGCCTCCTGTTTTGTTCTTCAATATTGTCT...  \n",
       "5   ACCTTAATTTTGACTTAGTCTAATTTAATCTTTTATGGTTAGTACT...  \n",
       "6   GAGTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAG...  \n",
       "7   CCCTTGATAGTGGAGACCATGTTTATACATTTTCATGGATCGAACA...  \n",
       "8   CCTTAAGGATTTTCTTCTCTGATTGCACTAAATCTATAGGTTTCTT...  \n",
       "9   GAGTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAG...  \n",
       "10  GAGTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAG...  \n",
       "11  CTTCGAGTCACTGGTTTGCCTGTATTGTGAAACCAACTGGATCCTG...  \n",
       "12  GTGGGTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAGGA...  \n",
       "13  CATGACACTGGCTTACCTTGTGCCAGGCAGATGGCAGCCACACAGT...  \n",
       "14  GTTTTATCAGCTTCCATACCCAAACAGAAATACCCTTAAGGATTTT...  \n",
       "15  CAGGCATTGAACAAGTACTAAGTGCCTGTTCTGTATCAGGTACTGT...  \n",
       "16  TCTGAATGTCATCGTCCACTCAGCCACTGGATTTAAGCAGAGTTCA...  \n",
       "17  AATACCCTTAAGGATTTTCTTCTCTGATTGCACTAAATCTATAGGT...  \n",
       "18  TATTGATGAAAGAAAGCTATTTCAGGAAACTTTTTTTTTTCTGTTA...  \n",
       "19  TAAGGATTTTCTTCTCTGATTGCACTAAATCTATAGGTTTCTTTGA...  \n",
       "20  TTGATGAAAGAAAGCTATTTCAGGAAACTTTTTTTTTTCTGTTAAT...  \n",
       "21  TTTTCTTCTCTGATTGCACTAAATCTATAGGTTTCTTTGATCCTAG...  \n",
       "22  TCAACTTTCAAAATACCTGCTGCTTCTTTCCAATTCACCTTTCTCT...  \n",
       "23  TCTCTGATTGCACTAAATCTATAGGTTTCTTTGATCCTAGAAATGA...  \n",
       "24  TCTGATTGCACTAAATCTATAGGTTTCTTTGATCCTAGAAATGAAT...  \n",
       "25  GCCTCCCTTTCCCGGGACAACAGAAGCTGACCTCTTTGATCTCTTG...  \n",
       "26  CAACATAGCCAACAAAAGGCTCTTTGTGTCTAATTTATATTCCTGA...  \n",
       "27  ATCCTAGAAATGAATAATAGAATATAAAGAATTCTTATTGATGAAA...  \n",
       "28  TAGAATATAAAGAATTCTTATTGATGAAAGAAAGCTATTTCAGGAA...  \n",
       "29  CCGGGACAACAGAAGCTGACCTCTTTGATCTCTTGCGCAGATGATG...  \n",
       "30  CCTCAGATGCTCTGTGCCTTGGATCTGGCCCCACTCCCGTCCTCCC...  \n",
       "31  AGTCTCCGGGGCTCTATGGGTTTCTGAATGTCATCGTCCACTCAGC...  \n",
       "32  TGGTTTGGGGAGGAGGGTTGCAGCGGCCGAGCCAGGGTCTCCACCC...  \n",
       "33  CGGGGCTCTATGGGTTTCTGAATGTCATCGTCCACTCAGCCACTGG...  \n",
       "34  TTCTGAATGTCATCGTCCACTCAGCCACTGGATTTAAGCAGAGTTC...  \n",
       "35  TTTTTGGCCCATGACACTGGCTTACCTTGTGCCAGGCAGATGGCAG...  \n",
       "36  CTCAGACACTGGATTTAAGCAGATTTCAAGTAAGTACTGGTTTGGG...  \n",
       "37  GTATGTTTTTAGCCAAAGACACTCGCTTACCTTGTGCCAGGCAGAT...  \n",
       "38  CACTGGATTTAAGCAGAGTTCAAGTAAGTACTGGTTTGGGGAGGAG...  \n",
       "39  CCTTATTACACTTCGAGTCACTGGTTTGCCTGTATTGTGAAACCAA...  \n",
       "40  TACTGGTTTGGGGAGGAGGGTTGCAGCGGCCGAGCCAGGGTCTCCA...  \n",
       "41  CCTCCCTTTCCCGGGACAACAGAAGCTGACCTCTTTGATCTCTTGC...  \n",
       "42  CACTGGTTTGCCTGTATTGTGAAACCAACTGGATCCTGAGATCCCC...  \n",
       "43  CTCCACCCAGGAAGGACTAATCGGGCAGGGTGTGGGGAAACAGGGA...  \n",
       "44  CCCTCTGCTGTCCTTGGAACCTTATTACACTTCGAGTCACTGGTTT...  \n",
       "45  CAGGAAGGACTAATCGGGCAGGGTGTGGGGAAACAGGGAGGTTGTT...  \n",
       "46  TCCCCAAGACAGAAATCATGATGAGTATGTTTTTGGCCCATGACAC...  \n",
       "47  AGGAAGGACTAATCGGGCAGGGTGTGGGGAAACAGGGAGGTTGTTC...  \n",
       "\n",
       "[48 rows x 22 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reads at junction defined by chr9 133607146 and chr22 23632741:\n",
      "Number split pairs: 5.0\n",
      "Number nonsplit pairs: 24.0\n",
      "Number paired: 58.0\n",
      "Total Number of split reads: 5.0\n",
      "Total Number of paired reads: 58.0\n"
     ]
    }
   ],
   "source": [
    "bamfile = \"/home/asureshkumar/Documents/research/ecDNA-Alignment/K562/K562_hg19_cnvkit.cs.rmdup.bam\"\n",
    "sumfile = \"/home/asureshkumar/Documents/research/ecDNA-Alignment/K562/K562_summaries.tsv\"\n",
    "samfile = pysam.AlignmentFile(bamfile, \"rb\")\n",
    "df = pd.read_csv(sumfile, sep=\"\\t\")\n",
    "# Define output table\n",
    "output = pd.DataFrame()\n",
    "num_split = 0\n",
    "num_paired = 0\n",
    "df = df.drop([\"pos1_flanking_coordinate\", \"pos2_flanking_coordinate\", \"sample\"], axis=1)\n",
    "display(df)\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "\n",
    "    # Fetch the DataFrames for split alignments and paired alignments\n",
    "    split_df, nonsplit_df = fetch_alignments(bamfile, *row, 150)\n",
    "    display(split_df)\n",
    "    display(nonsplit_df)\n",
    "    num_split += len(split_df)/3\n",
    "    num_paired += (2 * len(split_df) / 3) + len(nonsplit_df)\n",
    "    # num_paired += len(paired_df)\n",
    "    \n",
    "\n",
    "    print(f\"Reads at junction defined by {row.chrom1} {row.pos1} and {row.chrom2} {row.pos2}:\")\n",
    "    print(f\"Number split pairs: {len(split_df)/3}\")\n",
    "    print(f\"Number nonsplit pairs: {len(nonsplit_df)/2}\")\n",
    "    print(f\"Number paired: {(2 * len(split_df) / 3) + len(nonsplit_df)}\")\n",
    "    output = pd.concat([split_df, nonsplit_df], ignore_index=True)\n",
    "    break\n",
    "\n",
    "output.to_csv('alignments.tsv', sep='\\t', index=False)\n",
    "\n",
    "print(\"Total Number of split reads:\", num_split)\n",
    "print(\"Total Number of paired reads:\", num_paired)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
